packages/dht/src/connection/ConnectionManager.ts

```diff
@@ -0,0 +1,379 @@
+                    const connectivityResponseMessage: ConnectivityResponseMessage = {
+                        openInternet: true,
+                        ip: this.config.webSocketHost!,
+                        natType: 'open_internet',
+                        websocket: { ip: this.config.webSocketHost!, port: this.config.webSocketPort! }
+                    }
+                    resolve(connectivityResponseMessage)
+                }
+            })
+        }
+        const connectivityResponseMessage: ConnectivityResponseMessage = {
+            openInternet: false,
+            ip: 'localhost',
+            natType: 'unknown'
+        }
+        return connectivityResponseMessage
+    }
+
+    enableConnectivity(ownPeerDescriptor: PeerDescriptor): void {
+        if (!this.started || this.stopped) {
+            return
+        }
+        this.ownPeerDescriptor = ownPeerDescriptor
+        this.webSocketConnector!.setOwnPeerDescriptor(ownPeerDescriptor)
+        if (this.webSocketServer) {
+            this.webSocketServer.setOwnPeerDescriptor(ownPeerDescriptor)
+        }
+        if (this.webrtcConnector) {
+            this.webrtcConnector.setOwnPeerDescriptor(ownPeerDescriptor)
+            this.webrtcConnector.bindListeners(this.onIncomingMessage.bind(this), this.PROTOCOL_VERSION)
+        }
+
+        this.webSocketConnector!.bindListeners(this.onIncomingMessage.bind(this), this.PROTOCOL_VERSION)
+    }
+
+    onIncomingMessage = (connection: IConnection, message: Message): void => {
+        if (!this.started || this.stopped) {
+            return
+        }
+        if (message.messageType === MessageType.HANDSHAKE && this.ownPeerDescriptor) {
+            const handshake = HandshakeMessage.fromBinary(message.body)
+            const hexId = PeerID.fromValue(handshake.sourceId).toMapKey()
+            connection.setPeerDescriptor(handshake.peerDescriptor as PeerDescriptor)
+            if (!this.connections.hasOwnProperty(hexId)
+                || (this.connections[hexId] && this.connections[hexId].connectionType === ConnectionType.DEFERRED)) {
+                let oldConnection
+                if ((this.connections[hexId] && this.connections[hexId].connectionType === ConnectionType.DEFERRED)) {
+                    oldConnection = this.connections[hexId]
+                }
+                this.connections[hexId] = connection
+
+                const outgoingHandshake: HandshakeMessage = {
+                    sourceId: this.ownPeerDescriptor.peerId,
+                    protocolVersion: this.PROTOCOL_VERSION,
+                    peerDescriptor: this.ownPeerDescriptor
+                }
+                const msg: Message = {
+                    appId: this.CONNECTION_MANAGER_APP_ID,
+                    messageType: MessageType.HANDSHAKE, 
+                    messageId: v4(),
+                    body: HandshakeMessage.toBinary(outgoingHandshake)
+                }
+                connection.send(Message.toBinary(msg))
```

Comment:
Is it intentional that if we have a deferred connection, we first send the incoming message and then the buffered messages? Or should we send the buffered messages first?

Reply:
Yes, the first message that is sent here is a response to the handshake message. Anything in the buffer should be higher level stuff.